http/server.ts

```diff
@@ -1,111 +1,117 @@
 // Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
-import { listen, Conn, toAsyncIterator, Reader, Writer, copy } from "deno";
-import { BufReader, BufState, BufWriter } from "../io/bufio.ts";
+
+import { Conn, copy, listen, Reader, toAsyncIterator, Writer } from "deno";
+import { BufReader, BufWriter } from "../io/bufio.ts";
 import { TextProtoReader } from "../textproto/mod.ts";
 import { STATUS_TEXT } from "./http_status.ts";
 import { assert } from "../testing/mod.ts";
+import { defer, Deferred } from "../async/deferred.ts";
+import { Key, pathToRegexp } from "./path_to_regexp.ts";
+import { BodyReader, ChunkedBodyReader } from "./readers.ts";
 
-interface Deferred {
-  promise: Promise<{}>;
-  resolve: () => void;
-  reject: () => void;
-}
+export type HttpHandler = (req: ServerRequest) => Promise<any>;
 
-function deferred(): Deferred {
-  let resolve, reject;
-  const promise = new Promise((res, rej) => {
-    resolve = res;
-    reject = rej;
-  });
-  return {
-    promise,
-    resolve,
-    reject
-  };
-}
-
-interface ServeEnv {
-  reqQueue: ServerRequest[];
-  serveDeferred: Deferred;
-}
+export type ServerRequest = {
```

Comment:
Out of curiosity, what's the advantage of using interface over class?

Reply:
For test mocking.  In test codes, `HttpHandler` will be able to accept mocked request and dummy responder.